docs/proposals/protobuf.md

```diff
@@ -0,0 +1,501 @@
+resembles:
+
+          etcd -> decode -> defaulting -> convert to internal ->
+    JSON          50us      5us           15us
+    Proto         5us
+    JSON          150allocs               80allocs
+    Proto         100allocs
+
+          process -> convert to external -> encode -> client
+    JSON             15us                   40us
+    Proto                                   5us
+    JSON             80allocs               100allocs
+    Proto                                   4allocs
+
+ Protobuf has a huge benefit on encoding because it does not need to allocate
+ temporary objects, just one large buffer. Changing to protobuf moves our
+ hotspot back to conversion, not serialization.
+
+
+## Design Points
+
+* Generate Protobuf schema from Go structs (like we do for JSON) to avoid
+  manual schema update and drift
+* Generate Protobuf schema that is field equivalent to the JSON fields (no
+  special types or enumerations), reducing drift for clients across formats.
+* Follow our existing API versioning rules (backwards compatible in major
+  API versions, breaking changes across major versions) by creating one
+  Protobuf schema per API type.
+* Continue to use the existing REST API patterns but offer an alternative
+  serialization, which means existing client and server tooling can remain
+  the same while benefiting from faster decoding.
+* Protobuf objects on disk or in etcd will need to be self identifying at
+  rest, like JSON, in order for backwards compatibility in storage to work,
+  so we must add an envelope with apiVersion and kind to wrap the nested
+  object, and make the data format recognizable to clients.
+* Use the [gogo-protobuf](https://github.com/gogo/protobuf) Golang library to generate marshal/unmarshal
+  operations, allowing us to bypass the expensive reflection used by the
+  golang JSOn operation
+
+
+## Alternatives
+
+* We considered JSON compression to reduce size on wire, but that does not
+  reduce the amount of memory garbage created during serialization and
+  deserialization.
+* More efficient formats like Msgpack were considered, but they only offer
+  2x speed up vs the 10x observed for Protobuf
+* gRPC was considered, but is a larger change that requires more core
+  refactoring. This approach does not eliminate the possibility of switching
+  to gRPC in the future.
+* We considered attempting to improve JSON serialization, but the cost of
+  implementing a more efficient serializer library than ugorji is
+  significantly higher than creating a protobuf schema from our Go structs.
+
+
+## Schema
+
+The Protobuf schema for each API group and version will be generated from
+the objects in that API group and version. The schema will be named using
+the package identifier of the Go package, i.e.
+
+    k8s.io/kubernetes/pkg/api/v1
+
+Each top level object will be generated as a Protobuf message, i.e.:
+
+    type Pod struct { ... }
+
+    message Pod {}
```

Comment:
Just to clarify if I understand correctly:

"type Pod struct { ... }" - those are types that already exist, right?

"message Pod {}" - this is something that we want to auto-generate, right? 
And IIUC, it should also be: "message Pod { ... }" - because there should be the same fields, there. Or am I missing something?